### PR TITLE
Make use of `owningModel()` and `owningComponent()`

### PR DESCRIPTION
**Pull request by agarny**
Addresses issue #623.

### CODE REVIEW

**Comment by agarny** [comment]
@hsorby, @kerimoyle and @nickerso: this is a straightforward PR. Shouldn't take more than a minute to review.

**Review by kerimoyle — COMMENTED** [review]
I wonder if we can make life easier for users too by putting this functionality somewhere they can access it?  At present the user still has to do the casting themselves ... eg in this test:

```
TEST(Units, parentOfUnits)
{
    libcellml::ModelPtr model = libcellml::Model::create();
    model->setName("I_am_a_model");
    libcellml::UnitsPtr u = libcellml::Units::create();
    u->setName("u");
    model->addUnits(u);

    // Would be great to be able to say:
    // auto parent = owningModel(u); instead of the line below ... 
    libcellml::ModelPtr parent = std::dynamic_pointer_cast<libcellml::Model>(u->parent());
    EXPECT_FALSE(parent == nullptr);
    EXPECT_EQ(model, parent);
}
```

**Comment by agarny** [comment]
> I wonder if we can make life easier for users too by putting this functionality somewhere they can access it? At present the user still has to do the casting themselves ... eg in this test:

Yes, I thought the same when I came across that test. I guess this could be part of our super duper toolbox... :)

**Comment by agarny** [comment]
Ping @hsorby, @kerimoyle and @nickerso.

**Comment by kerimoyle** [comment]
Like I said earlier, I think that the `owningModel` and `owningComponent` functions should be made available to users too?  I wouldn't have thought that needed a separate PR?

**Comment by agarny** [comment]
You are asking for a coupe of methods to be added to the API. This PR is about cleaning up the code. So, two different things. Also, I don't think those methods should be part of the API, but rather of our "toolbox".

File: src/api/libcellml/units.h [hunk 1]
```diff
@@ -423,7 +431,7 @@ class LIBCELLML_EXPORT Units: public NamedEntity, public ImportedEntity
      * @return @c true when these @c Units rely on @c Units which are imported,
      * or @c false otherwise.
      */
-    bool requiresImports() const;
```
Comment (hsorby) [inline src/api/libcellml/units.h:426]: Did we have to drop the const here?

Comment (agarny) [reply src/api/libcellml/units.h:426]: Yes, because `Units::requiresImports()` now calls `shared_from_this()`.

**Comment by kerimoyle** [comment]
Hi again @agarny.  Could you add a catch for a `nullptr` entity argument please?  It will fix a bug I just found elsewhere :)

**Comment by agarny** [comment]
In `owningModel()` I imagine?

**Comment by kerimoyle** [comment]
Actually, in `owningComponent` ... there are spots where it takes an argument of `variable->parent()`, and if the parent is null then it segfaults.  Prob best to put it in both of them though?

**Comment by agarny** [comment]
Ok, I am going to do that check in both and return `nullptr` in such a case.

**Comment by agarny** [comment]
Ok, done in commit bb12545.

**Comment by agarny** [comment]
Ok, I am going to have to revert commit bb12545. Indeed, `owningModel()` and `owningComponent()` are utilities, so we can directly test them with `nullptr` as a parameter value. The only way would be to test them indirectly, but our current code only makes "proper" use of of those utilities, so...

Bottom line: `owningModel()` and `owningComponent()` should only be used when you are 100% certain that the parameter is not `nullptr`. After all, those are utilities, not methods that are part of the API (if they were, then it would be possible to test them).

**Comment by kerimoyle** [comment]
Do you mind if I change these functions in my PR then?  I found this while writing #630 so could get the coverage from undoing my checks there?

**Comment by agarny** [comment]
Where in your PR do you call `owningModel()` and/or `owningComponent()` in such a way that it segfaults? From what I can tell, all your calls to those methods look find (i.e. the entity that is passed is not `nullptr`).

Otherwise, to change those utilities in your PR won't solve the issue of the coverage...

**Comment by kerimoyle** [comment]
It (was) here: https://github.com/cellml/libcellml/pull/630/files#r440560834

**Review by kerimoyle — APPROVED** [review]

**Review by nickerso — APPROVED** [review]

**Pull request merged**